### ISSUE TITLE
Lookup types do not work well with type guards

### ISSUE THREAD

**Issue by mrhania** [Microsoft/TypeScript#12805]
**TypeScript Version:**  2.1.4

Consider this simple property getter function (defined in *What's New* of TypeScript 2.1):
```ts
function prop<T, K extends keyof T>(obj: T, key: K): T[K] {
    return obj[key];
}
```

I would expect that `prop(foo, "bar")` works the same way `foo.bar` does. However, if the type of the property is narrowed using type guards, the `prop` function stops working as expected:
 
```ts
interface Foo {
    bar?: number;
}

function quux(foo: Foo) {
    if (foo.bar) {
        let baz: number = prop(foo, "bar"); // error TS2322: Type 'number | undefined' is not assignable to type 'number'.
    }
}
```

**Expected behavior:** The code should compile just fine.
**Actual behavior:** Compilation ends with a type error.

**Comment by RyanCavanaugh** [comment]
Tentatively a duplicate of #10065 - type guards affect properties, but don't synthesize the kinds of on-the-fly new types that would be required to identify this in the general case. But here we *should* be ending up at the same symbol and resolving this correctly

**Comment by RyanCavanaugh** [comment]
@ahejlsberg any cheap fix possible?

**Comment by ahejlsberg** [comment]
@RyanCavanaugh Nothing cheap comes to mind. The real fix would be to address the issue in #10065 (which would fix several other issues as well).
